pkg/operator/ceph/csi/peermap/config.go

```diff
@@ -0,0 +1,309 @@
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package peermap
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"os"
+	"reflect"
+	"strconv"
+
+	"github.com/coreos/pkg/capnslog"
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/clusterd"
+	cephclient "github.com/rook/rook/pkg/daemon/ceph/client"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	"github.com/rook/rook/pkg/util"
+	"github.com/rook/rook/pkg/util/exec"
+	v1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	mappingConfigName = "rook-ceph-csi-mapping-config"
+	mappingConfigkey  = "csi-mapping-config-json"
+)
+
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", "peer-map")
+
+type PeerIDMapping struct {
+	ClusterIDMapping map[string]string
+	RBDPoolIDMapping []map[string]string
+}
+
+type PeerIDMappings []PeerIDMapping
+
+// addClusterIDMapping adds cluster ID map if not present already
+func (m *PeerIDMappings) addClusterIDMapping(newClusterIDMap map[string]string) {
+	if m.clusterIDMapIndex(newClusterIDMap) == -1 {
+		newDRMap := PeerIDMapping{
+			ClusterIDMapping: newClusterIDMap,
+		}
+		*m = append(*m, newDRMap)
+	}
+}
+
+// addRBDPoolIDMapping adds all the pool ID maps for a given cluster ID map
+func (m *PeerIDMappings) addRBDPoolIDMapping(clusterIDMap, newPoolIDMap map[string]string) {
+	for i := 0; i < len(*m); i++ {
+		if reflect.DeepEqual((*m)[i].ClusterIDMapping, clusterIDMap) {
+			(*m)[i].RBDPoolIDMapping = append((*m)[i].RBDPoolIDMapping, newPoolIDMap)
+		}
+	}
+}
+
+func (m *PeerIDMappings) clusterIDMapIndex(newClusterIDMap map[string]string) int {
+	for i, mapping := range *m {
+		if reflect.DeepEqual(mapping.ClusterIDMapping, newClusterIDMap) {
+			return i
+		}
+	}
+	return -1
+}
+
+func (m *PeerIDMappings) String() (string, error) {
+	mappingInBytes, err := json.Marshal(m)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to marshal peer cluster mapping config")
+	}
+
+	return string(mappingInBytes), nil
+}
+
+func toObj(in string) (PeerIDMappings, error) {
+	var mappings PeerIDMappings
+	err := json.Unmarshal([]byte(in), &mappings)
+	if err != nil {
+		return mappings, errors.Wrap(err, "failed to unmarshal peer cluster mapping config")
+	}
+
+	return mappings, nil
+}
+
+func ReconcilePoolIDMap(clusterContext *clusterd.Context, clusterInfo *cephclient.ClusterInfo) error {
+	pools := &cephv1.CephBlockPoolList{}
```

Comment:
Since this is triggered from the pool controller you will get the CephBlockPool object, so there is no need to list all the pools. Actually this will conflict with each reconcile.

So just pass CephBlockPool from the pool controller when calling this.